### DISCUSSION TITLE
2019 RogueLike Tutorial Port

### DISCUSSION THREAD

**Discussion by HexSleeves** [rust-gamedev/rust-game-ports#136] [Ideas]
Hey guys! I am currently working on porting Brackets 2019 rougelike tutorial to bevy. [Bevy Port](https://github.com/lecoqjacob/bevy_blood_oath).

I setup the effects system to work just like it does in the tutorial but I think I could be more bevy with it. Instead of a exclusive system that runs at the end of all the system stages, I was thinking of removing the queue completely and turning all the effects into events. What are y'all thoughts on this?

**Comment by 64kramsystem** [comment]
I'm not familiar neither with the source codebase, nor with Specs - the structure is considerably different from the book project.

Based on a cursory look of the source code, and the [related tutorial chapter](http://bfnightly.bracketproductions.com/rustbook/chapter_63.html?highlight=run_effects_queue#applying-damage), dispatchers need to be exclusive because multiple threads may run concurrently.

I'm not sure about the reason for the port to use a global for the effects queue, but globals of this type are modeled by resources (that is, you put the queue in a resource, and essentially it becomes an ECS-managed global).

An advantage of proper modeling, is that the ECS will manage concurrent access, and _won't_ schedule two systems concurrently if they both perform writes (AFAIK). Given this, you won't even need the world lock.

After this consideration, the next consideration is that yes, Bevy events may be the closest concept that model the effects queue - as a matter of fact, events are queues themselves :smile:. There's a caveat: I haven't had a look at the project design around queues, but Bevy events are consumed indipendently by events. Based on my cursory look at the source project, if a systems pops a queue element, the element will be permanently consumed; this is not the case in Bevy, where a given event is consumed per-system (as a matter of fact, in Bevy you don't "consume" events, you read them.

I think it's a good idea to ask to somebody in the Bevy discord server, also :smile:

**Reply by HexSleeves** [reply]
Using the queue as a resource never really occurred to me and honestly it could probably be a resource in the original project as well. I'm going to play with both implementations (events and resources) and see which I prefer. I'll branch off from the main source just so we can get a view of how both interact with each other

**Reply by HexSleeves** [reply]
After playing around a bit I was able to get single effects working (like consuming and item and that is it), but for effects that caused a cascade of more effects, I was unable to get working.

To explain this problem I will use the fireball scrol. When you trigger it, an effect is added to the queue and when consumed fires a item event. That item even handles if it is a consumable and if it applies damage, then it adds another effect for damage.

The problem is with events, it takes 1 frame for them to be picked up by other systems so my queue would trigger for the item event then when the damage event came in, other systems have already been triggered (like the movement system for the AI).

So whenever I use a targeting spell, the AI moves before all the events have finished coming in and being consumed. I don't actually know how to handle this :/

**Reply by 64kramsystem** [reply]
Somebody who was asking something similar on Reddit: https://www.reddit.com/r/rust_gamedev/comments/wgy3k8/how_can_i_loop_a_certain_stage_multiple_times_in.

Somebody on Bevy's Discord channel may give an alternate solution (or confirm that there isn't an "ideal" one).
